src/ray/core_worker/core_worker.cc

```diff
@@ -677,6 +685,63 @@ void CoreWorker::InternalHeartbeat(const boost::system::error_code &error) {
   internal_timer_.async_wait(boost::bind(&CoreWorker::InternalHeartbeat, this, _1));
 }
 
+void CoreWorker::LocationResolveHeartBeat(const boost::system::error_code &error) {
+  if (error == boost::asio::error::operation_aborted) {
+    return;
+  }
+  // Note: This will poll the GCS service every certain period.
+  // This exists to resolve scenarios such as
+  // https://github.com/ray-project/ray/pull/8679/files
+  ResolveActorsLocationNotPersistedToGCS();
+  internal_timer_.expires_at(
+      internal_timer_.expiry() +
+      boost::asio::chrono::milliseconds(kLocationResolveHeartbeatMillis));
+  internal_timer_.async_wait(
+      boost::bind(&CoreWorker::LocationResolveHeartBeat, this, _1));
+}
+
+void CoreWorker::ResolveActorsLocationNotPersistedToGCS() {
+  absl::MutexLock lock(&actor_location_resolve_waiters_mutex_);
+  for (auto it = actor_location_resolve_waiters_.begin();
+       it != actor_location_resolve_waiters_.end();) {
+    auto current = it++;
+    const auto &actor_id = *current;
+    ActorHandle *actor_handle = nullptr;
+    RAY_CHECK_OK(GetActorHandle(actor_id, &actor_handle));
+
+    if (actor_handle->IsPersistedToGCS()) {
+      // if an actor is already persisted to GCS, GCS will be responsible for reporting
+      // state of actors, so just remove it from the waiter.
+      actor_location_resolve_waiters_.erase(actor_id);
+    } else {
+      // If it is still not persisted to GCS, check worker/node is still alive.
+      // Disconnect actors if the worker or node is dead.
+      // This is needed because until the owner's local depedencies are resolved,
+      // actors won't be persisted to GCS, and there's no way we can know the states of
+      // them. Question? Is is possible worker is restarted after reporting failures?
+      const WorkerID &worker_id =
+          WorkerID::FromBinary(actor_handle->GetOwnerAddress().worker_id());
+      RAY_CHECK_OK(gcs_client_->Workers().AsyncGetWorkerFailure(
+          worker_id,
+          [this, actor_id](Status status,
+                           const boost::optional<gcs::WorkerFailureData> &result) {
+            if (status.ok() && result) {
+              direct_actor_submitter_->DisconnectActor(actor_id, true);
```

Comment:
There are two race conditions here that I think we need to handle:
1. Since GCS notifications are async, I think it's possible that the actor's location has just been registered, but then the owner dies concurrently, and we receive the lookup reply before the actor's location. I'm actually not sure if this can happen based on the GCS's current implementation, but I think it would be good in general not to depend on ordering between different GCS tables. So here, we should actually also do a lookup to the actor's entry to make sure it hasn't been added yet. The GCS service should also be checking whether the owner has already died when it receives an actor registration request so that it can drop the request, but you can leave that out for now (maybe add a TODO if you don't address in this PR).
2. We need to make sure that the GCS persisted flag is still unset in the actor handle when we disconnect the actor. Unfortunately, the mutex internal to the actor handle won't be enough since the caller can't hold the mutex while calling DisconnectActor.

Reply:
Addressed both!